### PR TITLE
Update functions.R

### PR DESCRIPTION
**Pull request by kraneguo**
Add a new example of clustered interference to the package (in the example part of our main function). Besides, some functions are also added at the bottom of this file to smoothly generate simulation data.

### CODE REVIEW

File: R/functions.R [hunk 1]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
   mean(Y[ind1])-mean(Y[ind2])
 }
 
+#' Define a house structure vector that describes the population in each household, which is useful in the clustered interference
```
Comment (dpuelz) [inline R/functions.R:369]: "Define a house structure vector that describes the population in each household, which is useful in the clustered interference.  In this running example, houses refer to clusters and kids refer to experimental units."

File: R/functions.R [hunk 2]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
...
+#'
+#' @return A list specifies treated houses and treated kids
+#' @export
+out_treat_household = function(housestruct, K1){
+  which_house = sample(1:length(housestruct),size=K1,replace=F) # which households to treat?
+  which_kid = apply(as.matrix(housestruct[which_house]),1,sample_mod) # which kids to treat (conditional on the treated households)?
+  return(list(which_house=which_house,which_kid=which_kid))
+}
+
+#' Gives a range of indexes that are exposed to treatments
```
Comment (dpuelz) [inline R/functions.R:411]: "Gives a range of indices that are exposed to treatments"

File: R/functions.R [hunk 3]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
...
+#'
+#' @param ii the index of our interest
```
Comment (dpuelz) [inline R/functions.R:413]: "ii the index of interest"

File: R/functions.R [hunk 4]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
...
+#' @param lind the first index in each household
+#' @param hind the last index in each household
+#'
+#' @return A vector of index exposed to the treament
```
Comment (dpuelz) [inline R/functions.R:417]: "A vector of indices exposed to the treatment"

File: R/functions.R [hunk 5]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
...
+  expind = c(apply(as.matrix(1:length(which_house)),1,out_exp_ind,lind=lind,hind=hind))
+  Z[expind,3] = 1
+  Z
+}
+
+#' Returns matrix \code{Z} of dimension (number of units x number of randomizations, i.e. assignments.)
+#'
+#' @param N The number of observations (kids).
+#' @param K The number of houses.
+#' @param equal A binary parameter that determines the exact house structure. If equal=TRUE, every household has equal number of kids. If equal=FALSE, every household samples the number of kids at random.
```
Comment (dpuelz) [inline R/functions.R:456]: "... every household is constructed by sampling the number of kids at random."

File: R/functions.R [hunk 6]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
...
+#' @param numrand The number of randomizations
+#' 
+#' @return A matrix compounds treated units as well as exposure functions. 0 represents a pure control unit. 1 represents a spillover unit. And 2 represents a treated unit.
```
Comment (dpuelz) [inline R/functions.R:459]: "A matrix of dimension (number of units x number of randomizations, i.e. assignments.) comprised of 0, denoting a unit is pure control under that particular assignment, 1 denoting a unit is a spillover under that particular assignment, and 2, denoting a unit is treated under that particular assignment."

File: R/functions.R [hunk 7]
```diff
@@ -343,3 +366,165 @@ ate = function(Z,Y){
...
+#' @param sig_c Standard error on the causal effect.
+#' @param sig_y Standard error on the outcome vector.
+#' @param taus Spillover effect.
+#' @param taup Primary effect.
+#' @param mu00 The effect on pure control units.
+#' @param equal A binary parameter that determines the exact house structure. If equal=TRUE, every household has equal number of kids. If equal=FALSE, every household samples the number of kids at random.
+#' 
+#' @return An outcome vector.
+#' @export
+out_bassefeller = function(N, K, Zobs, tau_main,
```
Comment (dpuelz) [inline R/functions.R:488]: maybe include the year of the basse / feller paper when citing, I think it's 2017?  Also, 

(1) You can remove the commented out lines of code in the function below if they are unnecessary
(2) Have you run the example to make sure everything works?

Comment (kraneguo) [reply R/functions.R:488]: Hi David, I have run the example when there is a positive spillover effect and the code returns a p value = 0, which from my prospective, rejects the null hypothesis of no spillover effect. I have also checked the validity under the null hypothesis and it's a valid test.

Comment (kraneguo) [reply R/functions.R:488]: The arxiv version of Basse, Feller's paper is in 2017, but the published version on JASA is in 2018. As a result, I add the year of the paper as 2018.

**Pull request merged**